src/Plugin/AddPathPlugin.php

```diff
@@ -32,17 +41,39 @@ public function __construct(UriInterface $uri)
         }
 
         $this->uri = $uri;
+
+        $resolver = new OptionsResolver();
+        $this->configureOptions($resolver);
+        $options = $resolver->resolve($config);
+
+        $this->smart = $options['smart'];
     }
 
     /**
      * {@inheritdoc}
      */
     public function handleRequest(RequestInterface $request, callable $next, callable $first)
     {
-        $request = $request->withUri($request->getUri()
-            ->withPath($this->uri->getPath().$request->getUri()->getPath())
-        );
+        $prepend = $this->uri->getPath();
+        $path = $request->getUri()->getPath();
+
+        if (!$this->smart || substr($path, 0, strlen($prepend)) !== $prepend) {
```

Comment:
This only consider the case where `$prepend` is the beginning of `$path`. Is that the only case we want? 

In your case on Gitlab, you do not want to prepend if a absolute URL is used. IMHO, that logic should only be in BaseUriPlugin.